include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -1324,6 +1324,292 @@ namespace Differentiation
+     *       for (unsigned int q_point = 0; q_point < n_q_points; ++q_point)
+     *       {
+     *         // Calculate the deformation gradient and magnetic field at this
+     *         // quadrature point
+     *         const Tensor<2, dim, ADNumberType> F =
+     *           unit_symmetric_tensor<dim>() + Grad_u[q_point];
+     *         const Tensor<1, dim, ADNumberType> H = -Grad_msp[q_point];
+     *         Assert(numbers::value_is_greater_than(determinant(F), 0.0),
+     *                ExcMessage("Negative jacobian detected!"));
+     *
+     *         // Extract some configuration dependent variables from our
+     *         // nonlinear constitutive law for the current quadrature point.
+     *         // In this way they only have to be computed once per quadrature
+     *         // point.
+     *         const SymmetricTensor<2,dim,ad_type> S = get_S(F,H);
+     *         const Tensor<1,dim,ad_type>          B = get_B(F,H);
+     *
+     *         // Define some position-dependent aliases, to make the assembly
+     *         // process easier to follow.
+     *         const double &JxW = fe_values.JxW(q_point);
+     *
+     *         // Add contribution of the internal forces:
+     *         // Note that we assemble the residual contribution directly
+     *         // as it is this vector that is to be automatically linearised.
+     *         for (unsigned int I = 0; I < n_dofs_per_cell; ++I)
+     *         {
+     *           // Determine the component and block associated with
+     *           // the I'th local degree-of-freedom.
+     *           const unsigned int block_I =
+     *             fe.system_to_base_index(I).first.first;
+     *
+     *           if (block_I == u_block) // u-terms
+     *           {
+     *             // Variation of the Green-Lagrange strain tensor
+     *             // associated with the I'th vector-valued basis function
+     *             const SymmetricTensor<2,dim,double> dE_I
+     *               = symmetrize(transpose(F)
+     *               * fe_values[u_fe].gradient(I,q_point));
+     *
+     *             residual_ad[I] += (dE_I*S) * JxW;
+     *           }
+     *           else if (block_I == msp_block)
+     *           {
+     *             // Variation of the magnetic field vector associated with
+     *             // the I'th scalar-valued basis function
+     *             const Tensor<1,dim,double> dH_I
+     *               = -fe_values[msp_fe].gradient(I, q_point);
+     *
+     *             residual_ad[I] -= (dH_I*B) * JxW;
+     *           }
+     *         }
+     *       }
+     *
+     *       // Add contribution from external sources. If these contributions
+     *       // are also solution dependent then they will also be consistently
+     *       // linearized.
+     *       // Loop over faces and accumulate external contributions into the
+     *       // cell total residual.
+     *       // residual_ad[I] += ...
```

Comment:
since it's all just pseudo-code, I would just show this as part of the regular code, not of the commentary